### PR TITLE
GPR review: Internalize most of gpr_thd (but not ID/current ID)

### PR DESCRIPTION
**Pull request by vjpai**
Part of grpc/proposal#56 . As an additional benefit, internalizing this will make it easier to change its API in the near future to make it more stable (with a 2-phase construction like all our other async things)

@apolcyn : Please review the change needed in C#. I also cleaned up a pre-existing use of an internal header.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%    +160 [None]    +712  +0.0%

  +0.0%    +160 TOTAL     +712  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%    +160 [None]    +712  +0.0%

  +0.0%    +160 TOTAL     +712  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by vjpai** [comment]
Ok, it turns out that this won't work for C# after all. The existing use of `src/core/lib/gpr` in C# is actually unnecessary and doesn't actually pick up any symbols. I'll respin this.

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%    +160 [None]    +752  +0.0%

  +0.0%    +160 TOTAL     +752  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: include/grpc/support/thd_id.h [hunk 1]
```diff
@@ -16,15 +16,27 @@
...
 
-#ifndef GRPC_CORE_LIB_GPR_THD_INTERNAL_H
-#define GRPC_CORE_LIB_GPR_THD_INTERNAL_H
+#ifndef GRPC_SUPPORT_THD_ID_H
+#define GRPC_SUPPORT_THD_ID_H
+/** Thread interface for GPR.
 
-#include <grpc/support/time.h>
+   Types
+        gpr_thd_id        a unique opaque identifier for a thread.
```
Comment (yashykt) [inline include/grpc/support/thd_id.h]: any special reason why we want to put this in a separate file?

Comment (vjpai) [reply include/grpc/support/thd_id.h]: I would have liked to fully internalize the whole thing, but wrapped languages need thread-id for logging purposes, and they have to get it from `include/grpc`, not from `src/lib`. So this stuff stays behind in public headers but all the other contents can get internalized. C# in particular is using this function and type.

Comment (yashykt) [reply include/grpc/support/thd_id.h]: ah Ok. got it. let's add this as a comment, so it's clear to future readers.

Comment (vjpai) [reply include/grpc/support/thd_id.h]: Done. Thanks!

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%    +128 [None]    +744  +0.0%

  +0.0%    +128 TOTAL     +744  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%    +128 [None]    +744  +0.0%

  +0.0%    +128 TOTAL     +744  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**